doc/man1/ecpg.1
.\\" auto-generated by docbook2man-spec $Revision: 1.25 $
.TH "ECPG" "1" "2002-11-22" "Application" "PostgreSQL Client Applications"
.SH NAME
ecpg \- embedded SQL C preprocessor
.SH SYNOPSIS
.sp
\fBecpg\fR\fR [ \fR\fB\fIoption\fB\fR...\fB \fR\fR]\fR \fB\fIfile\fB\fR...\fB\fR
.SH "DESCRIPTION"
.PP
\fBecpg\fR is the embedded SQL preprocessor for C
programs. It converts C programs with embedded SQL statements to
normal C code by replacing the SQL invocations with special
function calls. The output files can then be processed with any C
compiler tool chain.
.PP
\fBecpg\fR will convert each input file given on the
command line to the corresponding C output file. Input files
preferrably have the extension \fI.pgc\fR, in which
case the extension will be replaced by \fI.c\fR to
determine the output file name. If the extension of the input file
is not \fI.pgc\fR, then the output file name is
computed by appending .c to the full file name.
The output file name can also be overridden using the
\fB-o\fR option.
.PP
This reference page does not describe the embedded SQL language.
See the \fIPostgreSQL Programmer's Guide\fR for that.
.SH "OPTIONS"
.PP
\fBecpg\fR accepts the following command-line
arguments:
.TP
\fB-c\fR
Automatically generate C code from SQL code. Currently, this
works for EXEC SQL TYPE.
.TP
\fB-D \fIsymbol\fB\fR
Define a C preprocessor symbol.
.TP
\fB-I \fIdirectory\fB\fR
Specify an additional include path, used to find files included
via EXEC SQL INCLUDE. Defaults are
\fI\&.\fR (current directory),
\fI/usr/local/include\fR, the
PostgreSQL include directory which
is defined at compile time (default:
\fI/usr/local/pgsql/include\fR), and
\fI/usr/include\fR, in that order.
.TP
\fB-o \fIfilename\fB\fR
Specifies that \fBecpg\fR should write all
its output to the given \fIfilename\fR.
.TP
\fB-t\fR
Turn on autocommit of transactions. In this mode, each query is
automatically committed unless it is inside an explicit
transaction block. In the default mode, queries are committed
only when \fBEXEC SQL COMMIT\fR is issued.
.TP
\fB-v\fR
Print additional information including the version and the
include path.
.TP
\fB---help\fR
Show a brief summary of the command usage, then exit.
.TP
\fB--version\fR
Output version information, then exit.
.PP
.SH "NOTES"
.PP
When compiling the preprocessed C code files, the compiler needs to
be able to find the \fBECPG\fR header files in the
PostgreSQL include directory. Therefore, one might have to use the
\fB-I\fR option when invoking the compiler (e.g.,
-I/usr/local/pgsql/include).
.PP
Programs using C code with embedded SQL have to be linked against
the \fIlibecpg\fR library, for example using the
flags -L/usr/local/pgsql/lib -lecpg.
.PP
The value of either of these directories that is appropriate for
the installation can be found out using \fBpg_config\fR(1).
.SH "EXAMPLES"
.PP
If you have an embedded SQL C source file named
\fIprog1.pgc\fR, you can create an executable
program using the following sequence of commands:
.sp
.nf
ecpg prog1.pgc
cc -I/usr/local/pgsql/include -c prog1.c
cc -o prog1 prog1.o -L/usr/local/pgsql/lib -lecpg
.sp
.fi
.SH "SEE ALSO"
.PP
\fIPostgreSQL Programmer's Guide\fR for a more
detailed description of the embedded SQL interface


ChangeLog
2003-09-25  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/test/examples/tcqsanity/tcqsanity.c: It runs now - also easy
	to disable specific queries - just change the run_queries[] array.

	* src/backend/nodes/outfuncs.c: Added printing support for
	FStemOuterHashkeys and FSteMClauses

	* src/backend/executor/nodeSteM.c: Cleaned up debugging

	* src/backend/executor/nodeFSteM.c (FSteMGenerateOutputTuple): Fix
	for bug #169 - we omit to use the interested_queries list of the
	stemclauses to turn on the queriesCompleted bits for any queries
	that are _not_ part of the interested_queries and are a part of
	the usedByQueries.

	* src/backend/executor/nodeCQEddy.c (processOutputQ): Cleaned up
	the debugging print statements.

2003-07-16  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/backend/executor/execQual.c: Changed ExecEvalVar to use the
	tuple descriptor in the IHT in the CQ case.

2003-07-14  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/test/examples/tcqsanity/tcqsanity.old.c: Convert C++ comments to C

	* src/test/examples/tcq-sample/trafficincidents.c: Convert C++ comments to C

	* src/test/examples/geoserver/tracelog.c: Convert C++ comments to C

	* src/test/examples/geoserver/gettrace.c: Convert C++ comments to C

	* src/test/examples/demo/tcqsanity.c: Convert C++ comments to C

	* src/include/nodes/plannodes.h: Convert C++ comments to C

	* src/include/nodes/execnodes.h: Convert C++ comments to C

	* src/include/executor/nodeEddy.h: Convert C++ comments to C

	* src/include/executor/intermediateTuple.h: Convert C++ comments to C

	* src/backend/utils/misc/guc.c: Convert C++ comments to C

	* src/backend/telegraphcq/wrapch.c: Convert C++ comments to C

	* src/backend/telegraphcq/telegraphcq.c: Convert C++ comments to C

	* src/backend/tcop/postgres.c: Convert C++ comments to C

	* src/backend/rqueue/ingresq.c: Convert C++ comments to C

	* src/backend/postmaster/postmaster.c: Convert C++ comments to C

	* src/backend/parser/parse_wrapper.c: Convert C++ comments to C

	* src/backend/optimizer/plan/eddy.c: Convert C++ comments to C

	* src/backend/optimizer/plan/converttofsched.c: Convert C++ comments to C

	* src/backend/optimizer/plan/convertoeddy.c: Convert C++ comments to C

	* src/backend/executor/routingPolicy.c: Convert C++ comments to C

	* src/backend/executor/nodeSteMHashjoin.c: Convert C++ comments to C

	* src/backend/executor/nodeSteM.c: Convert C++ comments to C

	* src/backend/executor/nodeScanModule.c: Convert C++ comments to C

	* src/backend/executor/nodeSQEddy.c: Convert C++ comments to C

	* src/backend/executor/nodeIndexSteM.c: Convert C++ comments to C

	* src/backend/executor/nodeFluxCons.c: Convert C++ comments to C

	* src/backend/executor/nodeEddy.c: Convert C++ comments to C

	* src/backend/executor/nodeCQEddy.c: Convert C++ comments to C

	* src/backend/executor/execQual.c: Convert C++ comments to C

	* src/backend/executor/execProcnode.c: Convert C++ comments to C

	* src/backend/executor/execMain.c: Convert C++ comments to C

2003-07-09  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/backend/executor/routingPolicy.c: Cosmetic

	* src/backend/executor/nodeFluxProd_FT.c: setSocketOptions

	* src/backend/executor/nodeFluxProd.c: setSocketOptions

	* src/backend/executor/nodeFluxCons_FT.c: Need include file

	* src/backend/executor/nodeFluxCons.c: Need include file

	* src/backend/executor/nodeFluxAcceptor.c: setSocketOptions

	* src/backend/executor/nodeFSched.c: timersub doesn't run on Solaris

	* configure.in: Changed version

	* configure: Reran autoconf

	* src/backend/commands/tcqcommands.c: Cosmetic

	* src/backend/commands/tablecmds.c: Reinstate check for precisely
	one timestamp constraint in a stream.

	* src/backend/access/common/tupdesc.c: Cosmetic

2003-07-08  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/include/telegraphcq/wrapch.h: Add prototype for
	setSocketOptions

	* src/backend/telegraphcq/wrapch.c: Change setSocketOptions to
	non-static function

	* src/backend/parser/gram.y: Compilation error .. grrr

	* src/backend/executor/nodeFSched.c: Use setSocketOptions for
	portable setting of non-blocking IO

2003-06-20  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/backend/optimizer/plan/eddy.c: Purged commented code

	* src/backend/optimizer/plan/convertoeddy.c: Cleaned up
	superfluous error checking for filters. This has to be tested more.

	* src/backend/executor/nodeEddy.c: Added include file

2003-06-07  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/backend/optimizer/plan/eddy.c (add_planitem_to_eddy): Stupid
	bug ! We fell foul of the fact that we no longer compact sources
	and modules. So the Eddy's sources and modules array can keep
	growing. I changed the assertion that checks if numSources is <= 8
	to actually check the number of "active" sources (that is, those
	with eddy->sources[i] non-null). Same for modules. Two new
	functions get_num_active_sources/modules added. Needs to be
	investigated more. Dangerous for the demo env. where we keep
	adding and removing queries.
  
	* src/backend/executor/nodeEddy.c (_getNodeInformation): Fix crash
	when subjected to an expression that is not an operator (happens
	with incidents-table join !) - if unchecked will cause crash in
	the presence of the dyncatalog
 
2003-06-06  Sailesh Krishnamurthy  <sailesh@cs.berkeley.edu>

	* src/backend/optimizer/plan/eddy.c (sanity_check_rt_sources):
	Extra debug LOG code.

	* src/backend/executor/nodeFSteM.c (ProcessEndFSteM): Fix nasty
	bug - the sourceStems bitmap wasn't being cleared as appropriate.

	* src/backend/lib/bit.c (BitArrayPrintDebug): Protect ourself from
	broken debug code.

	* src/backend/executor/nodeCQEddy.c: Protect ourself from broken
	debug code.

	* src/backend/lib/bit.c: Stupid bug

	* src/backend/executor/processNode.c: Add argument to
	tcq_operators_add

	* src/backend/executor/nodeCQEddy.c (update_weights): Amol magic
	to ensure we "unlearn" routing soon.

src/interfaces/ecpg/preproc/ecpg.c
/* $Header: /project/eecs/db/cvsroot/postgres/src/interfaces/ecpg/preproc/ecpg.c,v 1.2 2004/03/24 08:11:30 chungwu Exp $ */

/* New main for ecpg, the PostgreSQL embedded SQL precompiler. */
/* (C) Michael Meskes <meskes@postgresql.org> Feb 5th, 1998 */
/* Placed under the same license as PostgresSQL */

#include "postgres_fe.h"

#include <unistd.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

extern int	optind;
extern char *optarg;

#include "extern.h"

int			ret_value = 0,
			autocommit = false,
			auto_create_c = false;
struct _include_path *include_paths = NULL;
struct cursor *cur = NULL;
struct typedefs *types = NULL;
struct _defines *defines = NULL;

static void
help(const char *progname)
{
	printf("%s is the PostgreSQL embedded SQL preprocessor for C programs.\n\n",
		   progname);
	printf("Usage:\n"
		   "  %s [OPTION]... FILE...\n\n",
		   progname);
	printf("Options:\n");
	printf("  -c             automatically generate C code from embedded SQL code;\n"
		   "                 currently this works for EXEC SQL TYPE\n");
#ifdef YYDEBUG
	printf("  -d             generate parser debug output\n");
#endif
	printf("  -D SYMBOL      define SYMBOL\n");
	printf("  -I DIRECTORY   search DIRECTORY for include files\n");
	printf("  -o OUTFILE     write result to OUTFILE\n");
	printf("  -t             turn on autocommit of transactions\n");
	printf("  --help         show this help, then exit\n");
	printf("  --version      output version information, then exit\n");
	printf("\nIf no output file is specified, the name is formed by adding .c to the\n"
		   "input file name, after stripping off .pgc if present.\n");
	printf("\nReport bugs to <pgsql-bugs@postgresql.org>.\n");
}

static void
add_include_path(char *path)
{
	struct _include_path *ip = include_paths;

	include_paths = mm_alloc(sizeof(struct _include_path));
	include_paths->path = path;
	include_paths->next = ip;

}

static void
add_preprocessor_define(char *define)
{
	struct _defines *pd = defines;

	defines = mm_alloc(sizeof(struct _defines));
	defines->old = strdup(define);
	defines->new = strdup("");
	defines->pertinent = true;
	defines->next = pd;
}

int
main(int argc, char *const argv[])
{
	int			fnr,
				c,
				verbose = false,
				out_option = 0;
	struct _include_path *ip;
	char	   *progname;

	if (!strrchr(argv[0], '/'))
		progname = argv[0];
	else
		progname = strrchr(argv[0], '/') + 1;

	if (argc > 1)
	{
		if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-?") == 0)
		{
			help(progname);
			exit(0);
		}
		else if (strcmp(argv[1], "--version") == 0)
		{
			printf("ecpg (PostgreSQL %s) %d.%d.%d\n", PG_VERSION,
				   MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL);
			exit(0);
		}
	}

	add_include_path("/usr/include");
	add_include_path(INCLUDE_PATH);
	add_include_path("/usr/local/include");
	add_include_path(".");

	while ((c = getopt(argc, argv, "vco:I:tD:d")) != -1)
	{
		switch (c)
		{
			case 'o':
				yyout = fopen(optarg, PG_BINARY_W);
				if (yyout == NULL)
					perror(optarg);
				else
					out_option = 1;
				break;
			case 'I':
				add_include_path(optarg);
				break;
			case 't':
				autocommit = true;
				break;
			case 'v':
				verbose = true;
				break;
			case 'c':
				auto_create_c = true;
				break;
			case 'D':
				add_preprocessor_define(optarg);
				break;
			case 'd':
#ifdef YYDEBUG
				yydebug = 1;
#else
				fprintf(stderr, "%s: parser debug support (-d) not available\n",
						progname);
#endif
				break;
			default:
				fprintf(stderr, "Try '%s --help' for more information.\n", argv[0]);
				return ILLEGAL_OPTION;
		}
	}

	if (verbose)
	{
		fprintf(stderr, "%s, the PostgreSQL embedded C preprocessor, version %d.%d.%d\n",
				progname, MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL);
		fprintf(stderr, "exec sql include ... search starts here:\n");
		for (ip = include_paths; ip != NULL; ip = ip->next)
			fprintf(stderr, " %s\n", ip->path);
		fprintf(stderr, "end of search list\n");
		return 0;
	}

	if (optind >= argc)			/* no files specified */
	{
		fprintf(stderr, "%s: no input files specified\n", progname);
		fprintf(stderr, "Try '%s --help' for more information.\n", argv[0]);
		return (ILLEGAL_OPTION);
	}
	else
	{
		/* after the options there must not be anything but filenames */
		for (fnr = optind; fnr < argc; fnr++)
		{
			char	   *output_filename = NULL,
					   *ptr2ext;

			input_filename = mm_alloc(strlen(argv[fnr]) + 5);

			strcpy(input_filename, argv[fnr]);

			/* take care of relative paths */
			ptr2ext = strrchr(input_filename, '/');
			ptr2ext = (ptr2ext ? strrchr(ptr2ext, '.') : strrchr(input_filename, '.'));

			/* no extension? */
			if (ptr2ext == NULL)
			{
				ptr2ext = input_filename + strlen(input_filename);

				/* no extension => add .pgc */
				ptr2ext[0] = '.';
				ptr2ext[1] = 'p';
				ptr2ext[2] = 'g';
				ptr2ext[3] = 'c';
				ptr2ext[4] = '\0';
			}

			if (out_option == 0)	/* calculate the output name */
			{
				output_filename = strdup(input_filename);

				ptr2ext = strrchr(output_filename, '.');
				/* make extension = .c */
				ptr2ext[1] = 'c';
				ptr2ext[2] = '\0';

				yyout = fopen(output_filename, PG_BINARY_W);
				if (yyout == NULL)
				{
					perror(output_filename);
					free(output_filename);
					free(input_filename);
					continue;
				}
			}

			yyin = fopen(input_filename, PG_BINARY_R);
			if (yyin == NULL)
				perror(argv[fnr]);
			else
			{
				struct cursor *ptr;
				struct _defines *defptr;
				struct typedefs *typeptr;

				/* remove old cursor definitions if any are still there */
				for (ptr = cur; ptr != NULL;)
				{
					struct cursor *this = ptr;
					struct arguments *l1,
							   *l2;

					free(ptr->command);
					free(ptr->connection);
					free(ptr->name);
					for (l1 = ptr->argsinsert; l1; l1 = l2)
					{
						l2 = l1->next;
						free(l1);
					}
					for (l1 = ptr->argsresult; l1; l1 = l2)
					{
						l2 = l1->next;
						free(l1);
					}
					ptr = ptr->next;
					free(this);
				}
				cur = NULL;

				/* remove non-pertinent old defines as well */
				while (defines && !defines->pertinent)
				{
					defptr = defines;
					defines = defines->next;

					free(defptr->new);
					free(defptr->old);
					free(defptr);
				}

				for (defptr = defines; defptr != NULL; defptr = defptr->next)
				{
					struct _defines *this = defptr->next;

					if (this && !this->pertinent)
					{
						defptr->next = this->next;

						free(this->new);
						free(this->old);
						free(this);
					}
				}

				/* and old typedefs */
				for (typeptr = types; typeptr != NULL;)
				{
					struct typedefs *this = typeptr;

					free(typeptr->name);
					ECPGfree_struct_member(typeptr->struct_member_list);
					free(typeptr->type);
					typeptr = typeptr->next;
					free(this);
				}
				types = NULL;

				/* initialize whenever structures */
				memset(&when_error, 0, sizeof(struct when));
				memset(&when_nf, 0, sizeof(struct when));
				memset(&when_warn, 0, sizeof(struct when));

				/* and structure member lists */
				memset(struct_member_list, 0, sizeof(struct_member_list));

				/* finally the actual connection */
				connection = NULL;

				/* initialize lex */
				lex_init();

				/* we need several includes */
				fprintf(yyout, "/* Processed by ecpg (%d.%d.%d) */\n/* These four include files are added by the preprocessor */\n#include <ecpgtype.h>\n#include <ecpglib.h>\n#include <ecpgerrno.h>\n#include <sqlca.h>\n#line 1 \"%s\"\n", MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL, input_filename);

				/* and parse the source */
				yyparse();

				if (yyin != NULL)
					fclose(yyin);
				if (out_option == 0)
					fclose(yyout);
			}

			if (output_filename)
				free(output_filename);

			free(input_filename);
		}
	}
	return ret_value;
}
